### PR TITLE
ref(workflows): consolidate workflows based on their purpose

### LINKED ISSUE #6166: actions: use a name convention for workflow files/names
## Motivation

We'd like to standardize the way the workflows are named, and align their names with what's actually being executed inside the workflow.

### Specifications

When we do the redesign of Zebra CI, can we use a naming standard for workflow files?
For example, we could name them something like: `(purpose)-(when they run)-(where they run)-(what they do)-(patch?).yml`

So we'd end up with names like:
- `sub-build-docker-image.yml`
- `ci-pr-gcp-continuous-integration-test-docker.yml`
- `ci-pr-gh-continuous-integration-test-os.yml`
- `cd-release-gcp-continuous-delivery-deploy-docker.yml`

### LINKED ISSUE #6167: actions: segregate workflows based on their purpose
## Motivation

Some workflows have different requirements like resources, time and dependencies, and even have jobs with different objectives (in some cases, not related to other workflow jobs).

### Specifications

If multiple workflows require the same setup, this should be split into a dependable workflow which sets all the information of the following workflows. And if this workflows are not dependable on each other or have different objectives, those should also live in their own workflow files.

Sometimes we require to restart a specific workflow, which also triggers a whole dependency tree which makes re-running time consuming and almost as similar as depending on all previous workflows if there's a false/positive.

### Complex Code or Requirements

- We can't make a dependable workflow depend on another dependable workflow
- Reduce workflow nesting as much as possible
- Be able to cancel specific time consuming jobs without affecting other jobs which ran successfully
- Resource intensive workflows/jobs should be split apart if other workflows don't have this resource heavy requirements

### PR DESCRIPTION
**Pull request by gustavovalverde**
## Motivation

Some workflows have different requirements like resources, time and dependencies, and even have jobs with different objectives (in some cases, not related to other workflow jobs).

Fixes: #6166
Fixes: #6167
Depends-On: #7660

### Complex Code or Requirements

To comply with the naming convention, some jobs had to be moved from one workflow to another, and we have a completely new workflow, but no new jobs, which consolidates integration tests (sync tests) which required to be deployed in GCP 

## Solution

- Rename the workflows to make their naming more consistent by adding a naming convention
- Move jobs to the workflow which is more consistent with their objective

## Review

- Confirm that all workflows are working as expected (running and not failing)
- Validate if the naming convention makes sense

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

### CODE REVIEW

**Comment by teor2345** [comment]
I really like the new names!
https://github.com/ZcashFoundation/zebra/tree/refactor-workflows/.github/workflows

But as I said in the meeting I am the worst person to review them, because I know what they do already.

**Comment by teor2345** [comment]
I think maybe this one is a typo?
https://github.com/ZcashFoundation/zebra/blob/refactor-workflows/.github/workflows/ci-unites-tests-os.patch.yml

The rest say `unit-tests`.

File: .github/workflows/ci-integration-tests-gcp.yml
Comment (teor2345) [inline .github/workflows/ci-integration-tests-gcp.yml:204]: Can you please run a manual full sync on this PR before merging it?
We'll need to fix bug #7618 to make full syncs work again.

(We forgot to run a full sync for the self-hosted runners change, and so I'm currently checking all the PRs back past that change as part of diagnosing bug #7618.)

File: .github/workflows/ci-unit-tests-docker.yml
Comment (teor2345) [inline .github/workflows/ci-unit-tests-docker.yml]: This job is timing out because it's configured for mainnet, not testnet. This seems like a bug in the workflow, because the Dockerfile hasn't changed.

**Review by teor2345 — DISMISSED** [review]
There's a bug in a testnet workflow that's making the tests fail.

File: .github/workflows/ci-integration-tests-gcp.yml
Comment (teor2345) [inline .github/workflows/ci-integration-tests-gcp.yml]: manual-find-cached-disks.yml is a sub-workflow, not a manual workflow.

Comment (gustavovalverde) [reply .github/workflows/ci-integration-tests-gcp.yml]: Fixed with [chore: use better name for find-disks](https://github.com/ZcashFoundation/zebra/pull/7616/commits/3db9e6c89733c5baf6c573a3d555a4f8c9178016)

File: .github/workflows/ci-unit-tests-docker.yml
Comment (gustavovalverde) [reply .github/workflows/ci-unit-tests-docker.yml]: This should fix it: [fix(ci): use the entrypoint.sh to change the Network](https://github.com/ZcashFoundation/zebra/pull/7616/commits/eac74b8e20f06826a546716f51044af111ae3dcd)

**Comment by gustavovalverde** [comment]
This one is ready @arya2

**Comment by gustavovalverde** [comment]
This should be ready

File: .github/workflows/cd-deploy-nodes-gcp.yml
Comment (teor2345) [inline .github/workflows/cd-deploy-nodes-gcp.yml:1]: Do we want to delay this large CI change until after the release?
I'll leave it to you and @upbqdn to decide, because he is the release manager.

Comment (gustavovalverde) [reply .github/workflows/cd-deploy-nodes-gcp.yml:1]: Same as in https://github.com/ZcashFoundation/zebra/pull/7682#discussion_r1357444420

Comment (gustavovalverde) [reply .github/workflows/cd-deploy-nodes-gcp.yml:1]: @upbqdn if by any means this delays you considerably for the release, just go ahead without these 2 PRs.

**Comment by teor2345** [comment]
It looks like this PR is missing some patch jobs, or they need to be renamed, or the branch protection rules need to be updated.

Different rules can be handled differently:
1. I don't think we need a `Run sync-to-checkpoint test` branch protection rule or patch job, because it almost never runs on PRs. (It doesn't even run in the scheduled job.)
2. The CD Docker rules are important for PRs, because we don't want to merge PRs that break Docker.

Screenshot 2023-10-13 at 13 25 27

File: .github/workflows/cd-deploy-nodes-gcp.yml
Comment (upbqdn) [reply .github/workflows/cd-deploy-nodes-gcp.yml:1]: This release will be big enough even without this change, so let's delay?

**Comment by teor2345** [comment]
> Different rules can be handled differently:
> 
> 1. I don't think we need a `Run sync-to-checkpoint test` branch protection rule or patch job, because it almost never runs on PRs. (It doesn't even run in the scheduled job.)

Admin: I have removed this useless branch protection rule, it's not actually checking anything in any PRs.

File: .github/workflows/cd-deploy-nodes-gcp.yml
Comment (teor2345) [inline .github/workflows/cd-deploy-nodes-gcp.yml]: This test is checking the production Docker image...

File: .github/workflows/ci-unit-tests-docker.yml
Comment (teor2345) [inline .github/workflows/ci-unit-tests-docker.yml]: But when it is moved to here, it is testing the test (CI) docker image.

Changing the image build will run the correct commands, because they now use the same `entrypoint.sh`. But they use different Docker scripts and Rust features, so we need these tests to run on the production Docker image.

(Otherwise we risk breaking the production image and not finding out until a user reports it.)

File: .github/workflows/ci-unit-tests-docker.yml
Comment (teor2345) [inline .github/workflows/ci-unit-tests-docker.yml]: This also applies to the testnet version of these tests.

File: .github/workflows/ci-unit-tests-docker.yml
Comment (teor2345) [inline .github/workflows/ci-unit-tests-docker.yml]: Even if it didn't before, this test should also be running on the production Docker image. We don't care how the test image handles custom configs, because we don't do configs that way in tests.

**Review by teor2345 — CHANGES_REQUESTED** [review]
I think there's an issue with the test coverage after this refactor. Some tests changed from running on the production image to the test image. But we need them to run on the production image, because that's the image that users use.

**Comment by teor2345** [comment]
> 2\. The CD Docker rules are important for PRs, because we don't want to merge PRs that break Docker.

These rules are not being satisfied because the new jobs are running on the wrong Docker images. Fixing that will make this PR merge.

**Comment by gustavovalverde** [comment]
> think there's an issue with the test coverage after this refactor. Some tests changed from running on the production image to the test image. But we need them to run on the production image, because that's the image that users use.

This has been fixed with:  [refactor: test config file in CI and CD with a reusable workflow](https://github.com/ZcashFoundation/zebra/pull/7616/commits/43ae06e7387d5b34b8c16f141b1017a595b5bd52)

**Comment by gustavovalverde** [comment]
This is ready to go!

File: .github/workflows/cd-deploy-nodes-gcp.yml
Comment (arya2) [inline .github/workflows/cd-deploy-nodes-gcp.yml]:
```suggestion
  # Test reconfiguring the docker image for testnet.
```

File: .github/workflows/ci-unit-tests-docker.yml
Comment (teor2345) [inline .github/workflows/ci-unit-tests-docker.yml]: Since these jobs run on both the CI (test) and CD (release) images, let's give them more specific names. The names also have to match the patch jobs.

```suggestion
    name: Test CI default Docker config file
```

File: .github/workflows/ci-unit-tests-docker.yml
Comment (teor2345) [inline .github/workflows/ci-unit-tests-docker.yml]:
```suggestion
    name: Test CI testnet Docker config file
```

File: .github/workflows/ci-unit-tests-docker.yml
Comment (teor2345) [inline .github/workflows/ci-unit-tests-docker.yml]:
```suggestion
    name: Test CI custom Docker config file
```

File: .github/workflows/ci-unit-tests-os.yml
Comment (teor2345) [inline .github/workflows/ci-unit-tests-os.yml]: The OS workflow does not use the Docker image, so it doesn't need to run when that file changes

Comment (teor2345) [reply .github/workflows/ci-unit-tests-os.yml]:
```suggestion
```

File: .github/workflows/ci-unit-tests-docker.patch.yml
Comment (teor2345) [inline .github/workflows/ci-unit-tests-docker.patch.yml]:
```suggestion
    name: Test CI default Docker config file
```

File: .github/workflows/ci-unit-tests-docker.patch.yml
Comment (teor2345) [inline .github/workflows/ci-unit-tests-docker.patch.yml]:
```suggestion
    name: Test CI custom Docker config file
```

File: .github/workflows/docs-deploy-firebase.patch.yml
Comment (teor2345) [inline .github/workflows/docs-deploy-firebase.patch.yml:21]: When we change the workflow file name it also needs to change in the patch workflow.

I applied this change to the branch because I can't comment on unchanged files.

File: .github/workflows/cd-deploy-nodes-gcp.yml
Comment (teor2345) [inline .github/workflows/cd-deploy-nodes-gcp.yml]: We can either change the branch protection rules to use the new name (I prefer this one for consistency)

```suggestion
    name: Test CD default Docker config file
```

Or change the job name to match the branch protection rules

```suggestion
    name: Test Zebra default Docker config file
```

File: .github/workflows/cd-deploy-nodes-gcp.yml
Comment (teor2345) [inline .github/workflows/cd-deploy-nodes-gcp.yml]: The production image name needs to be different from the CI (test) image jobs.

```suggestion
    name: Test CD testnet Docker config file
```

File: .github/workflows/cd-deploy-nodes-gcp.yml
Comment (teor2345) [inline .github/workflows/cd-deploy-nodes-gcp.yml]: We can either change the branch protection rules to use the new name (I prefer this one for consistency)

```suggestion
    name: Test CD custom Docker config file
```

Or change the job name to match the branch protection rules

```suggestion
    name: Test Zebra custom Docker config file
```

File: .github/workflows/cd-deploy-nodes-gcp.yml
Comment (teor2345) [inline .github/workflows/cd-deploy-nodes-gcp.yml:54]: The Docker release image checks need to run on PRs so that we don't merge PRs that break the release images.

File: .github/workflows/cd-deploy-nodes-gcp.yml
Comment (teor2345) [inline .github/workflows/cd-deploy-nodes-gcp.yml]: We can't depend on testnet jobs because testnet is unreliable.

```suggestion
    needs: [ build, versioning, test-configuration-file, test-zebra-conf-path ]
```

File: .github/workflows/cd-deploy-nodes-gcp.yml
Comment (teor2345) [inline .github/workflows/cd-deploy-nodes-gcp.yml]: This should depend on the 2 mainnet config file tests, because we don't want to deploy nodes if the image fails.

File: .github/workflows/continous-delivery.patch.yml
Comment (teor2345) [inline .github/workflows/continous-delivery.patch.yml:6]: This patch file is still required because we want the release image jobs to run on PRs. It will also need to be updated with the new set of jobs.

File: .github/workflows/ci-unit-tests-docker.yml
Comment (teor2345) [inline .github/workflows/ci-unit-tests-docker.yml]: The 2 new mainnet jobs in this file will also need new branch protection rules, because we don't want PRs to merge if the CI images stop working.

**Comment by teor2345** [comment]
We're almost there with the set of PR jobs and the branch protection rules, but we need to check the release images in every PR, just like we did before.

Otherwise we risk releasing an image that doesn't work.

**Comment by teor2345** [comment]
I think it's ok to change the branch protection rules now, as long as we get this PR in soon. But doing that will stop all the other PRs from merging.

**Comment by gustavovalverde** [comment]
**EDIT**: Actually, let's keep the test coverage as is, I'll make the workflow segregation in another PR [fix: revert test coverage on CD](https://github.com/ZcashFoundation/zebra/pull/7616/commits/100f95ed606afd1cdd485c86b78dc4ee1dabeba0)


> The Docker release image checks need to run on PRs so that we don't merge PRs that break the release images.

> we need to check the release images in every PR, just like we did before.

If we don't want to release a production image with issues, I think running this in the `main` branch should be a good starting point, as we'll be warned when this test fails.

A more "elegant" solution is also making the push on `Docker Hub` a dependency on this tests, at **release** time (but that's a refactor on its own). Otherwise, if we're going to test this in PRs I think this file should be segregated between a workflow that is meant to deploy instances (CD), and other that is meant to run tests on the test or release image (CI)

File: .github/workflows/cd-deploy-nodes-gcp.yml
Comment (gustavovalverde) [reply .github/workflows/cd-deploy-nodes-gcp.yml:54]: Reverted

File: .github/workflows/continous-delivery.patch.yml
Comment (gustavovalverde) [reply .github/workflows/continous-delivery.patch.yml:6]: Reverted

**Comment by gustavovalverde** [comment]
All recommendations were applied

**Comment by gustavovalverde** [comment]
I made a last fix to the docker config test to accept different images. Even though this was (mainly) required in this spots: https://github.com/ZcashFoundation/zebra/pull/7616/commits/fd6aaf217d5cce9b932cf8e8636e937e207a1700#diff-d43d0df5f20e7f637ae74715b01cc8a0b644574a546a5242c78b9b8d4bab9571R217

I added it to other places for consistency

**Comment by gustavovalverde** [comment]
If this test fails https://github.com/ZcashFoundation/zebra/actions/runs/6553940936/job/17800228563?pr=7616 after [fix(dockerfile): use variables or default for config path and file](https://github.com/ZcashFoundation/zebra/pull/7616/commits/3757299fb02209654c024f999eef07de14b60be6) it might because this file does not exists in the release image: `zebrad/tests/common/configs/v1.0.0-rc.2.toml`

If that's the case, we can remove this test from CD, or use a file generated by Zebra, different than the one created by the entrypoint

File: .github/workflows/cd-deploy-nodes-gcp.patch.yml
Comment (teor2345) [inline .github/workflows/cd-deploy-nodes-gcp.patch.yml]: Optional: testnet doesn't need a patch job, because we can't use it in branch protection rules, because it is unreliable

```suggestion
```

File: .github/workflows/cd-deploy-nodes-gcp.yml
Comment (teor2345) [inline .github/workflows/cd-deploy-nodes-gcp.yml:175]:
```suggestion
    needs: [ build, versioning, test-configuration-file, test-zebra-conf-path, test-configuration-file ]
```

File: .github/workflows/ci-unit-tests-docker.yml
Comment (teor2345) [inline .github/workflows/ci-unit-tests-docker.yml]: We don't want to open issues for testnet jobs, because they are unreliable

```suggestion
    needs: [ test-all, test-all-getblocktemplate-rpcs, test-fake-activation-heights, test-empty-sync, test-lightwalletd-integration, test-configuration-file, test-zebra-conf-path ]
```

**Comment by teor2345** [comment]
I'm going to delete the old branch protection rules now, so this PR and other PRs can merge.

Admin: This PR needs 4 new branch protection rules after it merges:
- [ ] Test CD default Docker config file / Test default-conf in Docker
- [ ] Test CD custom Docker config file / Test custom-conf in Docker
- [ ] Test CI default Docker config file / Test default-conf in Docker
- [ ] Test CI custom Docker config file / Test custom-conf in Docker

I'll go update the patch jobs to match in this PR now.

File: .github/workflows/cd-deploy-nodes-gcp.patch.yml
Comment (teor2345) [inline .github/workflows/cd-deploy-nodes-gcp.patch.yml]:
```suggestion
    name: Test CD default Docker config file / Test default-conf in Docker
```

File: .github/workflows/cd-deploy-nodes-gcp.patch.yml
Comment (teor2345) [inline .github/workflows/cd-deploy-nodes-gcp.patch.yml]:
```suggestion
    name: Test CD custom Docker config file / Test custom-conf in Docker
```

File: .github/workflows/ci-unit-tests-docker.patch.yml
Comment (teor2345) [inline .github/workflows/ci-unit-tests-docker.patch.yml]:
```suggestion
    name: Test CI default Docker config file / Test default-conf in Docker
```

File: .github/workflows/ci-unit-tests-docker.patch.yml
Comment (teor2345) [inline .github/workflows/ci-unit-tests-docker.patch.yml]:
```suggestion
    name: Test CI custom Docker config file / Test custom-conf in Docker
```

**Review by teor2345 — APPROVED** [review]
Thank you for all your work on this, I think we're good to go here

**Comment by teor2345** [comment]
> If we don't want to release a production image with issues, I think running this in the `main` branch should be a good starting point, as we'll be warned when this test fails.
> 
> A more "elegant" solution is also making the push on `Docker Hub` a dependency on this tests, at **release** time (but that's a refactor on its own).

We like to stop these kinds of bugs before we merge a PR, because fixing them after we merge or tag the release takes extra time. This is particularly important if we're trying to get a release out quickly to meet a deadline or fix a security issue.

It's ok to also test in the `main` branch if you want, but this seems less important (maybe optional?). It helps us work out if an issue is in multiple PRs, or the `main` branch and all PRs.

> Otherwise, if we're going to test this in PRs I think this file should be segregated between a workflow that is meant to deploy instances (CD), and other that is meant to run tests on the test or release image (CI)

That seems like a great idea! There's no need to have the release depend on the release image tests, as long as we're sure the tested release image is always the same as the DockerHub push released image. (If we're testing every PR, it's unlikely that the test will fail during the release itself.)

**Pull request merged**
